panel/template.py

```diff
@@ -0,0 +1,189 @@
+    else:
+        # Non-frozen Python and cx_Freeze can use __file__ directly
+        templates_path = osjoin(os.dirname(__file__), '_templates')
+
+    return Environment(loader=FileSystemLoader(templates_path))
+
+_env = get_env()
+
+
+class Template(object):
+    """
+    A Template is a high-level component to render multiple Panel
+    objects into a single HTML document. The Template object should be
+    given a string or jinja2 Template object in the constructor and
+    can then be populated with Panel objects. When adding panels to
+    the Template a unique name must be provided making it possible to
+    refer to them uniquely in the template e.g. two panels addded like
+    this:
+
+        template.add_panel('A', pn.panel('A'))
+        template.add_panel('B', pn.panel('B'))
+
+    May then be referenced in the template using the `embed` macro:
+
+        <div> {{ embed(roots.A) }} </div>
+        <div> {{ embed(roots.B) }} </div>
+
+    Once a template has been fully populated it can be rendered using
+    the same API as other Panel objects.
+    """
+
+    def __init__(self, template=None, items=None):
+        if isinstance(template, string_types):
+            template = _Template(template)
+        self.template = template
+        self._render_items = {}
+        items = {} if items is None else items
+        for name, item in items.items():
+            self.add_panel(name, item)
+
+    def _get_server(self, port=0, websocket_origin=None, loop=None,
+                   show=False, start=False, **kwargs):
+        return get_server(self, port, websocket_origin, loop, show,
+                          start, **kwargs)
+
+    def _modify_doc(self, server_id, doc):
+        """
+        Callback to handle FunctionHandler document creation.
+        """
+        if server_id:
+            state._servers[server_id][2].append(doc)
+        return self.server_doc(doc)
+
+    #----------------------------------------------------------------
+    # Public API
+    #----------------------------------------------------------------
+    
+    def add_panel(self, name, panel):
+        """
+        Add panels to the Template, which may then be referenced by
+        the given name using the jinja2 embed macro.
+
+        Arguments
+        ---------
+        name : str
+          The name to refer to the panel by in the template
+        panel : panel.Viewable
+          A Panel component to embed in the template.
+        """
+        
+        if name in self._render_items:
+            raise ValueError('The name %s has already been used for '
+                             'another panel. Ensure each panel '
+                             'has a unique name by which it can be '
+                             'referenced in the template.' % name)
+        self._render_items[name] = panel
+
+    def server_doc(self, doc=None, title=None):
+        """
+        Returns a serveable bokeh Document with the panel attached
```

Comment:
```suggestion
        Returns a servable bokeh Document with the panel attached
```

(to match spelling in the API)